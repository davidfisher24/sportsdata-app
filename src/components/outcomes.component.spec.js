import { render, cleanup } from '../test-utils'
import { Provider } from 'react-redux'
import Router, { BrowserRouter } from "react-router-dom";
import OutcomesList from './outcomes.component'
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useParams: jest.fn(),
}));

describe('Outcomes List Component', () => {
  let store; 

  afterEach(() => {
    cleanup() 
  })

  it('matches snapshot', () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ sportId: '1', eventId: '1' })
    store = mockStore({ 
        events: {
          data: [
            {id: 1, desc: 'Real Madrid vs Barcelona'},
            {id: 2, desc: 'Liverpool vs Arsenal'}
          ],
          error: null
        },
        outcomes: {
          data: [
            {cpid: 1, keyDimension: 'Draw', pr: '5/4', fdp: 1.2},
            {cpid: 2, keyDimension: 'Home', pr: '1/2', fdp: 0.5},
            {cpid: 3, keyDimension: 'Away', pr: '6/1', fdp: 6},
          ],
          error: null
        }
    });
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
            <OutcomesList />
        </BrowserRouter>
      </Provider>
    );
    expect(component.container).toMatchSnapshot()
  });

  it('renders correct event based on params', () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ sportId: '1', eventId: '1' })
    store = mockStore({ 
        events: {
          data: [
            {id: 1, desc: 'Real Madrid vs Barcelona'},
            {id: 2, desc: 'Liverpool vs Arsenal'}
          ],
          error: null
        },
        outcomes: {
          data: [
            {cpid: 1, keyDimension: 'Draw', pr: '5/4', fdp: 1.2},
            {cpid: 2, keyDimension: 'Home', pr: '1/2', fdp: 0.5},
            {cpid: 3, keyDimension: 'Away', pr: '6/1', fdp: 6},
          ],
          error: null
        }
    });
    const { getByTestId, getByText } = render(
      <Provider store={store}>
        <BrowserRouter>
            <OutcomesList />
        </BrowserRouter>
      </Provider>
    );
    expect(getByTestId("event-title")).toBeTruthy()
    expect(getByText(/- Real Madrid vs Barcelona/i)).toBeTruthy();
  });

  it('does not render event if not available based on params', () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ sportId: '1', eventId: '3' })
    store = mockStore({ 
        events: {
          data: [
            {id: 1, desc: 'Real Madrid vs Barcelona'},
            {id: 2, desc: 'Liverpool vs Arsenal'}
          ],
          error: null
        },
        outcomes: {
          data: [
            {cpid: 1, keyDimension: 'Draw', pr: '5/4', fdp: 1.2},
            {cpid: 2, keyDimension: 'Home', pr: '1/2', fdp: 0.5},
            {cpid: 3, keyDimension: 'Away', pr: '6/1', fdp: 6},
          ],
          error: null
        }
    });
    const { queryByTestId } = render(
      <Provider store={store}>
        <BrowserRouter>
            <OutcomesList />
        </BrowserRouter>
      </Provider>
    );
    expect(queryByTestId(/event-title/i)).toBeNull();
  });

  it('renders key dimension of event if available', () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ sportId: '1', eventId: '1' })
    store = mockStore({ 
        events: {
          data: [{id: 1, desc: 'Real Madrid vs Barcelona'},],
          error: null
        },
        outcomes: {
          data: [
            {cpid: 1, keyDimension: 'Draw', pr: '5/4', fdp: 1.2},
            {cpid: 2, keyDimension: 'Home', pr: '1/2', fdp: 0.5},
            {cpid: 3, keyDimension: 'Away', pr: '6/1', fdp: 6},
          ],
          error: null
        }
    });
    const { getAllByTestId } = render(
      <Provider store={store}>
        <BrowserRouter>
            <OutcomesList />
        </BrowserRouter>
      </Provider>
    );
    expect(getAllByTestId("key-dimension").length).toBe(3)
  });

  it('does not render key dimension of event if not available', () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ sportId: '1', eventId: '1' })
    store = mockStore({ 
        events: {
          data: [{id: 1, desc: 'Abu Dhabi Grand Prix'},],
          error: null
        },
        outcomes: {
          data: [
            {cpid: 1, d: 'Lewis Hamilton', pr: '5/4', fdp: 1.2},
            {cpid: 2, d: 'Max Verstappen', pr: '1/2', fdp: 0.5},
            {cpid: 3, d: 'Fernando Alonso', pr: '6/1', fdp: 6},
          ],
          error: null
        }
    });
    const { queryByTestId } = render(
      <Provider store={store}>
        <BrowserRouter>
            <OutcomesList />
        </BrowserRouter>
      </Provider>
    );
    expect(queryByTestId(/key-dimension/i)).toBeNull();
  });

  it('renders error on error data', () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ sportId: '1', eventId: '1' })
    store = mockStore({ 
        outcomes: {
          data: null,
          error: new Error
        },
        events: {
          data: [],
          error: null
        }
    });
    const { getByTestId } = render(
      <Provider store={store}>
        <BrowserRouter>
            <OutcomesList />
        </BrowserRouter>
      </Provider>
    );
    expect(getByTestId("error-container")).toBeTruthy();
  });
});