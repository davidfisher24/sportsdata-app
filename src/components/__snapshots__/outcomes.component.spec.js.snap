// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outcomes List Component matches snapshot 1`] = `
<div>
  <div>
    <div
      class="list row"
    >
      <div
        class="col-md-8"
      >
        <h1
          class="h4"
        >
          Outcomes List
          <span
            data-testid="event-title"
          >
             - 
            Real Madrid vs Barcelona
          </span>
        </h1>
        <ul
          class="list-group"
        >
          <div
            class="list-group-item list-group-item-action flex-column align-items-start"
          >
            <div
              class="d-flex w-100 justify-content-between"
            >
              <h5
                class="mb-1"
              />
              <small>
                5/4
              </small>
            </div>
            <p
              class="mb-1"
              data-testid="key-dimension"
            >
              Draw
            </p>
            <small>
              5/4
               - 
              1.2
            </small>
          </div>
          <div
            class="list-group-item list-group-item-action flex-column align-items-start"
          >
            <div
              class="d-flex w-100 justify-content-between"
            >
              <h5
                class="mb-1"
              />
              <small>
                1/2
              </small>
            </div>
            <p
              class="mb-1"
              data-testid="key-dimension"
            >
              Home
            </p>
            <small>
              1/2
               - 
              0.5
            </small>
          </div>
          <div
            class="list-group-item list-group-item-action flex-column align-items-start"
          >
            <div
              class="d-flex w-100 justify-content-between"
            >
              <h5
                class="mb-1"
              />
              <small>
                6/1
              </small>
            </div>
            <p
              class="mb-1"
              data-testid="key-dimension"
            >
              Away
            </p>
            <small>
              6/1
               - 
              6
            </small>
          </div>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
